serve/mlc_serve/model/paged_cache_model.py

```diff
@@ -104,6 +105,19 @@ def _is_safe_to_sample(prob_like):
             do_top_p |= top_ps[-1] < 1.0
             do_top_k |= top_ks[-1] != vocab_size
 
+            if not param.presence_penalty == 0.0 or not param.frequency_penalty == 0 and bool(freq):
+                freq_tensor = np.array(list(freq.items()))
+                index = torch.from_numpy(freq_tensor[..., 0]).to(device=logits.device)
+                src = torch.from_numpy(freq_tensor[..., 1]).type_as(logits).to(device=logits.device)
+                logits[i] = torch.scatter_add(logits[i], dim=0, index=index, src=-(src * param.frequency_penalty + param.presence_penalty))
+            
+            if param.logit_bias:
+                bias_tensor = np.array(list(param.logit_bias.items()))
+                index = torch.from_numpy(bias_tensor[..., 0]).to(device=logits.device)
+                src = torch.from_numpy(bias_tensor[..., 1]).type_as(logits).to(device=logits.device)
+                logits[i] = torch.scatter_add(logits[i], dim=0, index=index, src=src)
```

Comment:
Do we need this `scatter_add`? Can't we do something like
```
logits[i][index] += src
```

Since `logit_bias` is fixed, we can maintain a list without dup entries, can't we?

Reply:
that makes sense

Reply:
fixed